File: backend/services/agent_llm.py
"""Resolve the selected Agent LLM provider and run one tool-calling turn."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, cast
from uuid import uuid4

from api_types import AgentMessagePayload, AgentToolDeclarationPayload
from _routes._errors import HTTPError
from services.anthropic_messages_client import (
    anthropic_messages_from_agent,
    anthropic_tools_from_declarations,
    call_anthropic_messages_turn,
)
from services.gemini_text_client import (
    apply_gemini_thinking_config,
    call_gemini_generate_content_turn,
    resolve_gemini_model,
)
from services.interfaces import HTTPClient, JSONValue
from services.openai_compatible_chat import (
    call_openai_compatible_turn,
    openai_messages_from_agent,
    openai_tools_from_declarations,
)
from state.app_settings import (
    BUILTIN_GEMINI_PROVIDER_ID,
    CUSTOM_AGENT_LLM_KINDS,
    AgentLlmAuthMode,
    AgentLlmProviderKind,
    AgentLlmProviderSettings,
    AppSettings,
    provider_has_oauth,
    selected_agent_llm_provider,
)

AgentLlmApiKind = Literal["gemini", "openai", "anthropic"]


@dataclass(frozen=True, slots=True)
class AgentLlmCatalogEntry:
    kind: AgentLlmProviderKind
    label: str
    api_kind: AgentLlmApiKind
    default_model: str
    default_base_url: str
    key_url: str
    supports_oauth: bool = False
    oauth_api_kind: AgentLlmApiKind | None = None
    oauth_base_url: str = ""


AGENT_LLM_CATALOG: tuple[AgentLlmCatalogEntry, ...] = (
    AgentLlmCatalogEntry(
        kind="gemini",
        label="Gemini",
        api_kind="gemini",
        default_model="",
        default_base_url="https://generativelanguage.googleapis.com/v1beta",
        key_url="https://aistudio.google.com/app/apikey",
    ),
    AgentLlmCatalogEntry(
        kind="openai",
        label="OpenAI (ChatGPT)",
        api_kind="openai",
        default_model="gpt-4o",
        default_base_url="https://api.openai.com/v1",
        key_url="https://platform.openai.com/api-keys",
        supports_oauth=True,
        oauth_api_kind="openai",
        oauth_base_url="https://chatgpt.com/backend-api/codex",
    ),
    AgentLlmCatalogEntry(
        kind="anthropic",
        label="Anthropic",
        api_kind="anthropic",
        default_model="claude-sonnet-4-5",
        default_base_url="https://api.anthropic.com",
        key_url="https://console.anthropic.com/settings/keys",
        supports_oauth=True,
    ),
    AgentLlmCatalogEntry(
        kind="openrouter",
        label="OpenRouter",
        api_kind="openai",
        default_model="openai/gpt-4o",
        default_base_url="https://openrouter.ai/api/v1",
        key_url="https://openrouter.ai/keys",
    ),
    AgentLlmCatalogEntry(
        kind="zai",
        label="Z.ai",
        api_kind="openai",
        default_model="glm-4.5",
        default_base_url="https://api.z.ai/api/paas/v4",
        key_url="https://z.ai/manage-apikey/apikeys",
    ),
    AgentLlmCatalogEntry(
        kind="minimax",
        label="MiniMax",
        api_kind="openai",
        default_model="MiniMax-M2",
        default_base_url="https://api.minimax.io/v1",
        key_url="https://platform.minimax.io/user-center/basic-information/interface-key",
        supports_oauth=True,
        oauth_api_kind="anthropic",
        oauth_base_url="https://api.minimax.io/anthropic",
    ),
    AgentLlmCatalogEntry(
        kind="moonshot",
        label="Moonshot / Kimi",
        api_kind="openai",
        default_model="kimi-k2-0905-preview",
        default_base_url="https://api.moonshot.ai/v1",
        key_url="https://platform.moonshot.ai/console/api-keys",
        supports_oauth=True,
        oauth_base_url="https://api.kimi.com/coding/v1",
    ),
    AgentLlmCatalogEntry(
        kind="xai",
        label="xAI (Grok)",
        api_kind="openai",
        default_model="grok-4",
        default_base_url="https://api.x.ai/v1",
        key_url="https://console.x.ai/team/default/api-keys",
        supports_oauth=True,
    ),
    AgentLlmCatalogEntry(
        kind="groq",
        label="Groq",
        api_kind="openai",
        default_model="llama-3.3-70b-versatile",
        default_base_url="https://api.groq.com/openai/v1",
        key_url="https://console.groq.com/keys",
    ),
    AgentLlmCatalogEntry(
        kind="deepseek",
        label="DeepSeek",
        api_kind="openai",
        default_model="deepseek-chat",
        default_base_url="https://api.deepseek.com",
        key_url="https://platform.deepseek.com/api_keys",
    ),
    AgentLlmCatalogEntry(
        kind="custom_openai",
        label="Custom (OpenAI-compatible)",
        api_kind="openai",
        default_model="",
        default_base_url="",
        key_url="",
    ),
    AgentLlmCatalogEntry(
        kind="custom_anthropic",
        label="Custom (Anthropic-compatible)",
        api_kind="anthropic",
        default_model="",
        default_base_url="",
        key_url="",
    ),
)

_CATALOG_BY_KIND = {entry.kind: entry for entry in AGENT_LLM_CATALOG}


@dataclass(frozen=True, slots=True)
class LlmFunctionCall:
    name: str
    args: dict[str, JSONValue]
    call_id: str = ""


@dataclass(frozen=True, slots=True)
class LlmTurnResult:
    text: str
    function_calls: tuple[LlmFunctionCall, ...]


@dataclass(frozen=True, slots=True)
class ResolvedAgentLlm:
    provider_id: str
    kind: AgentLlmProviderKind
    api_kind: AgentLlmApiKind
    label: str
    api_key: str
    model: str
    base_url: str
    auth_mode: AgentLlmAuthMode = "api_key"
    oauth_account_id: str = ""
    oauth_refresh_token: str = ""
    oauth_expires_at: float = 0.0


def catalog_entry(kind: AgentLlmProviderKind) -> AgentLlmCatalogEntry:
    return _CATALOG_BY_KIND[kind]


def _json_object(value: object) -> dict[str, JSONValue]:
    if not isinstance(value, dict):
        return {}
    result: dict[str, JSONValue] = {}
    for key, item in cast(dict[object, object], value).items():
        if isinstance(key, str):
            result[key] = cast(JSONValue, item)
    return result


def contents_from_messages(messages: list[AgentMessagePayload]) -> list[JSONValue]:
    contents: list[JSONValue] = []
    for message in messages:
        parts: list[JSONValue] = []
        if message.role == "assistant":
            for part in message.parts:
                if part.type == "text" and part.text:
                    parts.append({"text": part.text})
                elif part.type == "tool_call" and part.toolName:
                    parts.append(
                        {
                            "functionCall": {
                                "name": part.toolName,
                                "args": _json_object(part.arguments),
                            }
                        }
                    )
            if parts:
                contents.append({"role": "model", "parts": parts})
            continue

        for part in message.parts:
            if part.type == "text" and part.text:
                parts.append({"text": part.text})
            elif part.type == "inline_image" and part.data and part.mimeType:
                parts.append({"inlineData": {"mimeType": part.mimeType, "data": part.data}})
            elif part.type == "tool_result" and part.toolName:
                response_payload = part.result if part.result is not None else {"ok": True}
                parts.append(
                    {
                        "functionResponse": {
                            "name": part.toolName,
                            "response": _json_object(response_payload),
                        }
                    }
                )
        if parts:
            contents.append({"role": "user", "parts": parts})
    if not contents:
        raise HTTPError(400, "Agent turn has no usable message parts")
    return contents


def resolve_agent_llm(settings: AppSettings, model_override: str | None = None) -> ResolvedAgentLlm:
    override = (model_override or "").strip()
    provider = selected_agent_llm_provider(settings)
    if provider is None:
        return _resolve_builtin_gemini(settings, override)

    entry = catalog_entry(provider.kind)
    use_oauth = provider_has_oauth(provider)
    api_key = provider.oauth_access_token if use_oauth else provider.api_key
    if provider.kind == "gemini" and not api_key:
        api_key = settings.gemini_api_key
    model = override or provider.model or entry.default_model
    if provider.kind == "gemini" and not model:
        model = settings.gemini_model
    api_kind = entry.oauth_api_kind if use_oauth and entry.oauth_api_kind else entry.api_kind
    base_url = provider.base_url or ((entry.oauth_base_url if use_oauth else "") or entry.default_base_url)
    if provider.kind in CUSTOM_AGENT_LLM_KINDS and (not base_url or not (override or provider.model)):
        raise HTTPError(400, "AGENT_LLM_PROVIDER_INVALID")
    return ResolvedAgentLlm(
        provider_id=provider.id,
        kind=provider.kind,
        api_kind=api_kind,
        label=provider.label or entry.label,
        api_key=api_key,
        model=model,
        base_url=base_url,
        auth_mode="oauth" if use_oauth else "api_key",
        oauth_account_id=provider.oauth_account_id,
        oauth_refresh_token=provider.oauth_refresh_token,
        oauth_expires_at=provider.oauth_expires_at,
    )


def _resolve_builtin_gemini(settings: AppSettings, model_override: str) -> ResolvedAgentLlm:
    entry = catalog_entry("gemini")
    return ResolvedAgentLlm(
        provider_id=BUILTIN_GEMINI_PROVIDER_ID,
        kind="gemini",
        api_kind="gemini",
        label=entry.label,
        api_key=settings.gemini_api_key,
        model=model_override or settings.gemini_model,
        base_url=entry.default_base_url,
    )


def missing_key_code(resolved: ResolvedAgentLlm) -> str:
    if resolved.kind == "gemini":
        return "GEMINI_API_KEY_MISSING"
    return "AGENT_LLM_KEY_MISSING"


def run_agent_llm_turn(
    http: HTTPClient,
    resolved: ResolvedAgentLlm,
    *,
    messages: list[AgentMessagePayload],
    available_tools: list[AgentToolDeclarationPayload],
    system_instruction: str,
) -> LlmTurnResult:
    if not resolved.api_key:
        raise HTTPError(400, missing_key_code(resolved))

    if resolved.api_kind == "gemini":
        gemini_model = resolve_gemini_model(resolved.model)
        result = call_gemini_generate_content_turn(
            http,
            api_key=resolved.api_key,
            model=gemini_model,
            contents=contents_from_messages(messages),
            system_instruction=system_instruction,
            tools=_gemini_tools(available_tools),
            generation_config=apply_gemini_thinking_config(
                gemini_model,
                {"temperature": 0.4, "maxOutputTokens": 2048},
            ),
            timeout=60,
        )
        return LlmTurnResult(
            text=result.text,
            function_calls=tuple(
                LlmFunctionCall(name=call.name, args=call.args, call_id=f"call_{uuid4().hex[:10]}")
                for call in result.function_calls
            ),
        )

    if resolved.kind == "openai" and resolved.auth_mode == "oauth":
        from services.codex_responses_client import call_codex_responses_turn

        return call_codex_responses_turn(
            http,
            access_token=resolved.api_key,
            account_id=resolved.oauth_account_id,
            model=resolved.model,
            messages=messages,
            available_tools=available_tools,
            system_instruction=system_instruction,
            timeout=60,
        )

    if resolved.api_kind == "openai":
        return call_openai_compatible_turn(
            http,
            api_key=resolved.api_key,
            model=resolved.model,
            base_url=resolved.base_url,
            messages=openai_messages_from_agent(messages, system_instruction),
            tools=openai_tools_from_declarations(available_tools),
            extra_headers=_openai_extra_headers(resolved.kind),
            timeout=60,
        )

    return call_anthropic_messages_turn(
        http,
        api_key=resolved.api_key,
        model=resolved.model,
        base_url=resolved.base_url,
        system_instruction=system_instruction,
        messages=anthropic_messages_from_agent(messages),
        tools=anthropic_tools_from_declarations(available_tools),
        use_oauth_bearer=resolved.auth_mode == "oauth",
        timeout=60,
    )


def _gemini_tools(declarations: list[AgentToolDeclarationPayload]) -> list[JSONValue] | None:
    if not declarations:
        return None
    function_declarations: list[JSONValue] = []
    for item in declarations:
        name = item.name.strip()
        if not name:
            continue
        function_declarations.append(
            {
                "name": name,
                "description": item.description,
                "parameters": _json_object(item.parameters) or {"type": "object", "properties": {}},
            }
        )
    if not function_declarations:
        return None
    return [{"functionDeclarations": function_declarations}]


def _openai_extra_headers(kind: AgentLlmProviderKind) -> dict[str, str] | None:
    if kind != "openrouter":
        return None
    return {
        "HTTP-Referer": "https://ltx.studio",
        "X-Title": "LTX Desktop",
    }


def merge_agent_llm_providers(
    existing: list[AgentLlmProviderSettings],
    incoming: list[object],
) -> list[JSONValue]:
    existing_by_id = {provider.id: provider for provider in existing}
    merged: list[JSONValue] = []
    seen_ids: set[str] = set()
    for item in incoming:
        if not isinstance(item, dict):
            continue
        fields = cast(dict[str, object], item)
        provider_id = str(fields.get("id") or "").strip()
        kind = fields.get("kind")
        if not provider_id or not isinstance(kind, str) or kind not in _CATALOG_BY_KIND:
            continue
        if provider_id in seen_ids:
            continue
        seen_ids.add(provider_id)
        previous = existing_by_id.get(provider_id)
        api_key = str(fields.get("api_key") or "").strip()
        if not api_key and previous is not None:
            api_key = previous.api_key
        oauth_access = str(fields.get("oauth_access_token") or "").strip()
        oauth_refresh = str(fields.get("oauth_refresh_token") or "").strip()
        oauth_account_id = str(fields.get("oauth_account_id") or "").strip()
        oauth_account_label = str(fields.get("oauth_account_label") or "").strip()
        raw_expires = fields.get("oauth_expires_at")
        oauth_expires_at = float(raw_expires) if isinstance(raw_expires, (int, float)) else 0.0
        auth_mode = fields.get("auth_mode")
        if auth_mode not in ("api_key", "oauth"):
            auth_mode = previous.auth_mode if previous is not None else "api_key"
        if previous is not None:
            if not oauth_access:
                oauth_access = previous.oauth_access_token
            if not oauth_refresh:
                oauth_refresh = previous.oauth_refresh_token
            if oauth_expires_at <= 0:
                oauth_expires_at = previous.oauth_expires_at
            if not oauth_account_id:
                oauth_account_id = previous.oauth_account_id
            if not oauth_account_label:
                oauth_account_label = previous.oauth_account_label
        if oauth_access and auth_mode != "api_key":
            auth_mode = "oauth"
        merged.append(
            {
                "id": provider_id,
                "kind": kind,
                "label": str(fields.get("label") or "").strip(),
                "api_key": api_key,
                "model": str(fields.get("model") or "").strip(),
                "base_url": str(fields.get("base_url") or "").strip(),
                "auth_mode": auth_mode,
                "oauth_access_token": oauth_access,
                "oauth_refresh_token": oauth_refresh,
                "oauth_expires_at": oauth_expires_at,
                "oauth_account_id": oauth_account_id,
                "oauth_account_label": oauth_account_label,
            }
        )
    return merged

File: frontend/lib/agent-llm.ts
export const BUILTIN_GEMINI_PROVIDER_ID = 'gemini'

export type AgentLlmProviderKind =
  | 'gemini'
  | 'openai'
  | 'anthropic'
  | 'openrouter'
  | 'zai'
  | 'minimax'
  | 'moonshot'
  | 'xai'
  | 'groq'
  | 'deepseek'
  | 'custom_openai'
  | 'custom_anthropic'

export type AgentLlmAuthMode = 'api_key' | 'oauth'

export interface AgentLlmProviderPublic {
  id: string
  kind: AgentLlmProviderKind
  label: string
  hasApiKey: boolean
  hasOAuth: boolean
  authMode: AgentLlmAuthMode
  model: string
  baseUrl: string
}

export interface AgentLlmCatalogEntry {
  kind: AgentLlmProviderKind
  label: string
  defaultModel: string
  defaultBaseUrl: string
  keyUrl: string
  requiresBaseUrl: boolean
  supportsConnect: boolean
}

export const AGENT_LLM_CATALOG: readonly AgentLlmCatalogEntry[] = [
  {
    kind: 'gemini',
    label: 'Gemini',
    defaultModel: '',
    defaultBaseUrl: '',
    keyUrl: 'https://aistudio.google.com/app/apikey',
    requiresBaseUrl: false,
    supportsConnect: false,
  },
  {
    kind: 'openai',
    label: 'OpenAI (ChatGPT)',
    defaultModel: 'gpt-4o',
    defaultBaseUrl: 'https://api.openai.com/v1',
    keyUrl: 'https://platform.openai.com/api-keys',
    requiresBaseUrl: false,
    supportsConnect: true,
  },
  {
    kind: 'anthropic',
    label: 'Anthropic',
    defaultModel: 'claude-sonnet-4-5',
    defaultBaseUrl: 'https://api.anthropic.com',
    keyUrl: 'https://console.anthropic.com/settings/keys',
    requiresBaseUrl: false,
    supportsConnect: true,
  },
  {
    kind: 'openrouter',
    label: 'OpenRouter',
    defaultModel: 'openai/gpt-4o',
    defaultBaseUrl: 'https://openrouter.ai/api/v1',
    keyUrl: 'https://openrouter.ai/keys',
    requiresBaseUrl: false,
    supportsConnect: false,
  },
  {
    kind: 'zai',
    label: 'Z.ai',
    defaultModel: 'glm-4.5',
    defaultBaseUrl: 'https://api.z.ai/api/paas/v4',
    keyUrl: 'https://z.ai/manage-apikey/apikeys',
    requiresBaseUrl: false,
    supportsConnect: false,
  },
  {
    kind: 'minimax',
    label: 'MiniMax',
    defaultModel: 'MiniMax-M2',
    defaultBaseUrl: 'https://api.minimax.io/v1',
    keyUrl: 'https://platform.minimax.io/user-center/basic-information/interface-key',
    requiresBaseUrl: false,
    supportsConnect: true,
  },
  {
    kind: 'moonshot',
    label: 'Moonshot / Kimi',
    defaultModel: 'kimi-k2-0905-preview',
    defaultBaseUrl: 'https://api.moonshot.ai/v1',
    keyUrl: 'https://platform.moonshot.ai/console/api-keys',
    requiresBaseUrl: false,
    supportsConnect: true,
  },
  {
    kind: 'xai',
    label: 'xAI (Grok)',
    defaultModel: 'grok-4',
    defaultBaseUrl: 'https://api.x.ai/v1',
    keyUrl: 'https://console.x.ai/team/default/api-keys',
    requiresBaseUrl: false,
    supportsConnect: true,
  },
  {
    kind: 'groq',
    label: 'Groq',
    defaultModel: 'llama-3.3-70b-versatile',
    defaultBaseUrl: 'https://api.groq.com/openai/v1',
    keyUrl: 'https://console.groq.com/keys',
    requiresBaseUrl: false,
    supportsConnect: false,
  },
  {
    kind: 'deepseek',
    label: 'DeepSeek',
    defaultModel: 'deepseek-chat',
    defaultBaseUrl: 'https://api.deepseek.com',
    keyUrl: 'https://platform.deepseek.com/api_keys',
    requiresBaseUrl: false,
    supportsConnect: false,
  },
  {
    kind: 'custom_openai',
    label: 'Custom (OpenAI-compatible)',
    defaultModel: '',
    defaultBaseUrl: '',
    keyUrl: '',
    requiresBaseUrl: true,
    supportsConnect: false,
  },
  {
    kind: 'custom_anthropic',
    label: 'Custom (Anthropic-compatible)',
    defaultModel: '',
    defaultBaseUrl: '',
    keyUrl: '',
    requiresBaseUrl: true,
    supportsConnect: false,
  },
]

export interface AgentLlmModelOption {
  id: string
  displayName: string
  description?: string
}

export const AGENT_LLM_CATALOG_MODELS: Record<AgentLlmProviderKind, readonly AgentLlmModelOption[]> = {
  gemini: [
    { id: 'gemini-3.5-flash-lite', displayName: 'Gemini 3.5 Flash-Lite' },
    { id: 'gemini-3.5-flash', displayName: 'Gemini 3.5 Flash' },
    { id: 'gemini-3-pro-preview', displayName: 'Gemini 3 Pro' },
    { id: 'gemini-2.5-flash', displayName: 'Gemini 2.5 Flash' },
    { id: 'gemini-2.5-pro', displayName: 'Gemini 2.5 Pro' },
    { id: 'gemini-2.0-flash', displayName: 'Gemini 2.0 Flash' },
  ],
  openai: [
    { id: 'gpt-5.4', displayName: 'GPT-5.4' },
    { id: 'gpt-5.3', displayName: 'GPT-5.3' },
    { id: 'gpt-5.2', displayName: 'GPT-5.2' },
    { id: 'gpt-5.1', displayName: 'GPT-5.1' },
    { id: 'gpt-5', displayName: 'GPT-5' },
    { id: 'gpt-4.1', displayName: 'GPT-4.1' },
    { id: 'gpt-4o', displayName: 'GPT-4o' },
    { id: 'gpt-4o-mini', displayName: 'GPT-4o mini' },
    { id: 'o3', displayName: 'o3' },
    { id: 'o4-mini', displayName: 'o4-mini' },
  ],
  anthropic: [
    { id: 'claude-opus-4-6', displayName: 'Claude Opus 4.6' },
    { id: 'claude-sonnet-4-6', displayName: 'Claude Sonnet 4.6' },
    { id: 'claude-opus-4-5', displayName: 'Claude Opus 4.5' },
    { id: 'claude-sonnet-4-5', displayName: 'Claude Sonnet 4.5' },
    { id: 'claude-haiku-4-5', displayName: 'Claude Haiku 4.5' },
    { id: 'claude-opus-4-1', displayName: 'Claude Opus 4.1' },
    { id: 'claude-sonnet-4', displayName: 'Claude Sonnet 4' },
  ],
  openrouter: [
    { id: 'openai/gpt-5', displayName: 'OpenAI: GPT-5' },
    { id: 'anthropic/claude-sonnet-4.5', displayName: 'Anthropic: Claude Sonnet 4.5' },
    { id: 'google/gemini-2.5-pro', displayName: 'Google: Gemini 2.5 Pro' },
    { id: 'openai/gpt-4o', displayName: 'OpenAI: GPT-4o' },
    { id: 'anthropic/claude-sonnet-4', displayName: 'Anthropic: Claude Sonnet 4' },
  ],
  zai: [
    { id: 'glm-4.6', displayName: 'GLM-4.6' },
    { id: 'glm-4.5', displayName: 'GLM-4.5' },
    { id: 'glm-4.5-air', displayName: 'GLM-4.5 Air' },
  ],
  minimax: [
    { id: 'MiniMax-M2.5', displayName: 'MiniMax M2.5' },
    { id: 'MiniMax-M2', displayName: 'MiniMax M2' },
    { id: 'MiniMax-M1', displayName: 'MiniMax M1' },
  ],
  moonshot: [
    { id: 'kimi-k2.5', displayName: 'Kimi K2.5' },
    { id: 'kimi-k2-0905-preview', displayName: 'Kimi K2 0905' },
    { id: 'kimi-k2-turbo-preview', displayName: 'Kimi K2 Turbo' },
    { id: 'moonshot-v1-128k', displayName: 'Moonshot v1 128k' },
  ],
  xai: [
    { id: 'grok-4', displayName: 'Grok 4' },
    { id: 'grok-3', displayName: 'Grok 3' },
    { id: 'grok-3-mini', displayName: 'Grok 3 Mini' },
    { id: 'grok-2', displayName: 'Grok 2' },
  ],
  groq: [
    { id: 'llama-3.3-70b-versatile', displayName: 'Llama 3.3 70B' },
    { id: 'openai/gpt-oss-120b', displayName: 'GPT-OSS 120B' },
    { id: 'meta-llama/llama-4-maverick-17b-128e-instruct', displayName: 'Llama 4 Maverick' },
    { id: 'qwen/qwen3-32b', displayName: 'Qwen 3 32B' },
  ],
  deepseek: [
    { id: 'deepseek-chat', displayName: 'DeepSeek Chat' },
    { id: 'deepseek-reasoner', displayName: 'DeepSeek Reasoner' },
  ],
  custom_openai: [],
  custom_anthropic: [],
}

export const CUSTOM_AGENT_LLM_MODEL_VALUE = '__custom__'

export function catalogModels(kind: AgentLlmProviderKind): AgentLlmModelOption[] {
  return [...(AGENT_LLM_CATALOG_MODELS[kind] ?? [])]
}

export function mergeAgentLlmModelOptions(
  fetched: readonly AgentLlmModelOption[],
  kind: AgentLlmProviderKind,
  includeId = '',
): AgentLlmModelOption[] {
  const ordered: AgentLlmModelOption[] = []
  const seen = new Set<string>()
  for (const model of [...fetched, ...catalogModels(kind)]) {
    const id = model.id.trim()
    if (!id || seen.has(id)) continue
    ordered.push(model)
    seen.add(id)
  }
  const included = includeId.trim()
  if (included && !seen.has(included)) {
    ordered.push({ id: included, displayName: included })
  }
  return ordered
}

export const AGENT_LLM_KEY_REQUIRED_SETTINGS_DETAIL = {
  tab: 'apiKeys' as const,
  reason: 'agentLlmKeyRequired' as const,
}

const AGENT_KEY_ERROR_CODES = new Set([
  'GEMINI_API_KEY_MISSING',
  'GEMINI_INVALID_API_KEY',
  'AGENT_LLM_KEY_MISSING',
  'AGENT_LLM_INVALID_API_KEY',
  'AGENT_LLM_PROVIDER_INVALID',
])

export function catalogEntry(kind: AgentLlmProviderKind): AgentLlmCatalogEntry {
  return AGENT_LLM_CATALOG.find(entry => entry.kind === kind) ?? AGENT_LLM_CATALOG[0]
}

export function providerDisplayLabel(provider: Pick<AgentLlmProviderPublic, 'kind' | 'label'>): string {
  return provider.label.trim() || catalogEntry(provider.kind).label
}

export function providerHasCredential(provider: AgentLlmProviderPublic): boolean {
  return provider.hasOAuth || provider.hasApiKey || provider.kind === 'gemini'
}

export function isAgentLlmKeyError(code?: string): boolean {
  return Boolean(code && AGENT_KEY_ERROR_CODES.has(code))
}

export function createAgentLlmProviderId(): string {
  if (typeof crypto !== 'undefined' && 'randomUUID' in crypto) {
    return `prov_${crypto.randomUUID()}`
  }
  return `prov_${Math.random().toString(36).slice(2, 12)}`
}
